### ISSUE TITLE
Widget Thumbnails

### ISSUE THREAD

**Issue by quadcom** [MilhouseVH/texturecache.py#67]
Thanks for building this little nugget. I seem to have discovered a situation that is perplexing.

I've run the script on a box that I moved over to a shared MySQL. Copied Texture13.db and replaced all the files in "Thumbnails" from a properly functioning system.

What I've run into is that while both movies and TV shows do display the correct thumbs, widgets don't. That is unless I go into each object and manually reselect the artwork for both poster and fanart. Once I do that then the widget images start to show. But I can't see myself doing that for over 2k objects!!

I had widgets for W1=In-Progress Movies and W2=New Added on the movies page (homescreen) and W1=New TV Shows on the TV page. 

./texturecache.py c @xbmc.host=192.168.200.197 @download.threads=10

Any ideas on what might be going on here? It's as if the widgets have a completely separate cache.

**Comment by quadcom** [comment]
Ran another 'c' pass. There are lots of skipped and duplicate in the report at the end. Not sure what I need to do from here.

Tried doing a purgetest (./texturecache.py purgetest hashed @xbmc.host=192.168.200.197) but either there was nothing or the script failed to run. 

```
adrian@mariadb:~$ ./texturecache.py c @xbmc.host=192.168.200.197 @webserver.port=8080
Need to cache: [thumbnail ] for addon: Default Controller
Need to cache: [  fanart  ] for addon: Aeon Nox: SiLVO
Need to cache: [thumbnail ] for addon: Aeon Nox: SiLVO
Need to cache: [thumbnail ] for addon: Elementtree XML parser
Need to cache: [  fanart  ] for addon: Picture Slideshow Screensaver
Need to cache: [thumbnail ] for addon: Picture Slideshow Screensaver
Need to cache: [  fanart  ] for addon: Zach Morris Add-ons
Need to cache: [thumbnail ] for addon: Zach Morris Add-ons
Need to cache: [  fanart  ] for addon: Gamestarter add-on repository
Need to cache: [thumbnail ] for addon: Gamestarter add-on repository
Need to cache: [thumbnail ] for addon: a4kSubtitles
Need to cache: [thumbnail ] for addon: NordVPN
Need to cache: [thumbnail ] for addon: PlayScrapers Module
Need to cache: [  fanart  ] for addon: PressPlay Artwork
Need to cache: [thumbnail ] for addon: PressPlay Artwork
Need to cache: [  fanart  ] for addon: PressPlay Metadata
Need to cache: [thumbnail ] for addon: PressPlay Metadata
Need to cache: [  fanart  ] for addon: [COLOR gold]PressPlay Repository[/COLOR]
Need to cache: [thumbnail ] for addon: [COLOR gold]PressPlay Repository[/COLOR]
Need to cache: [  fanart  ] for addon: [COLOR pink]ExodusVPN.com[/COLOR]
Need to cache: [thumbnail ] for addon: [COLOR pink]ExodusVPN.com[/COLOR]
Need to cache: [  fanart  ] for addon: Raspberry Pi VNC
Need to cache: [thumbnail ] for addon: Raspberry Pi VNC

Caching artwork: 0 items remaining of 23 (qs: 0, qm: 0), 0 errors, 0 threads active (01.00 downloads per second, ETA: 00:00:00)
Need to cache: [set.fanart] for movie: A Close Shave
Need to cache: [set.fanart] for movie: The Woman in Black

Caching artwork: 0 items remaining of 2 (qs: 0, qm: 0), 2 errors, 0 threads active (02.00 downloads per second, ETA: 00:00:00)

The following items could not be downloaded:

[set.fanart] [A Close Shave                           ] http://image.tmdb.org/t/p/original/fsv2eKzskPsEag3NORLQBW1wac5.jpg
[set.fanart] [The Woman in Black                      ] http://image.tmdb.org/t/p/original/pw8JrqNceLQ82KWZpHg5VghWTqn.jpg
Need to cache: [  fanart  ] for set: The Woman in Black Collection
Need to cache: [  fanart  ] for set: Wallace & Gromit Collection

Caching artwork: 0 items remaining of 2 (qs: 0, qm: 0), 2 errors, 0 threads active (02.00 downloads per second, ETA: 00:00:00)

The following items could not be downloaded:

[fanart    ] [Wallace & Gromit Collection             ] http://image.tmdb.org/t/p/original/fsv2eKzskPsEag3NORLQBW1wac5.jpg
[fanart    ] [The Woman in Black Collection           ] http://image.tmdb.org/t/p/original/pw8JrqNceLQ82KWZpHg5VghWTqn.jpg
Parsing Tvshows: Murder and Justice: The Case of Martha Moxley, Murder and Justice: The Case of Martha Moxley","next_episode_to_air":null,"networks":[],"number_of_episodes":3,"number_of_seasons":1,"origin_country":[],"original_language":"en","original_name":"Murder and Justice: The Case of Martha Moxley","overview":"This documentary profiles the 1975 murder of 15 year old Martha Moxley in the rich and well to do gated community of Greenwich Village, CT. Moxely was brutally beaten to death with a golf club on the eve of Halloween in 1975.","popularity":0.6,"poster_path":"/VGtoGuOsPZpOGYgECfIYEZGSP7.jpg","production_coParsing Tvshows: Murder and Justice: The Case of Martha Moxley, Murder and Justice: The Case of Martha Moxley","next_episode_to_air":null,"networks":[],"number_of_episodes":3,"number_of_seasons":1,"origin_country":[],"original_language":"en","original_name":"Murder and Justice: The Case of Martha Moxley","overview":"This documentary profiles the 1975 murder of 15 year old Martha Moxley in the rich and well to do gated community of Greenwich Village, CT. Moxely was brutally beaten to death with a golf club on the eve of Halloween in 1975.","popularity":0.6,"poster_path":"/VGtoGuOsPZpOGYgECfIYEZGSP7.jpg","production_coParsing Tvshows: Murder Mountain, Season 1...                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   Need to cache: [  poster  ] for tvshow: Be Cool, Scooby-Doo!, Season 1
Need to cache: [  thumb   ] for tvshow: The Flintstones, Season 4 Episode 4x21
Need to cache: [  poster  ] for tvshow: The World Wars, Season 1
Need to cache: [  thumb   ] for tvshow: This Is Us, Season 5 Episode 5x07

Caching artwork: 0 items remaining of 4 (qs: 0, qm: 0), 3 errors, 0 threads active (00.00 downloads per second, ETA: 00:00:00)

The following items could not be downloaded:

[poster    ] [Be Cool, Scooby-Doo!, Season 1          ] http://image.tmdb.org/t/p/original/im1zk7CNHXd1h172b10Z2Nug5UR.jpg
[poster    ] [The World Wars, Season 1                ] https://image.tmdb.org/t/p/original/zoHWpRN29rlmguiIQl4y7TTW19I.jpg
[thumb     ] [The Flintstones, Season 4 Episode 4x21  ] https://image.tmdb.org/t/p/original/bycnXEq3WrYXcPr7pgjGkpKVvuN.jpg

Cache pre-load activity summary for "addons/agenres/vgenres/albums/artists/movies/sets/tvshows":

              |    banner   |    fanart   |    poster   | season.banner | season.poster |  set.fanart |  set.poster |    thumb    |  thumbnail  |    TOTAL
--------------+-------------+-------------+-------------+---------------+---------------+-------------+-------------+-------------+-------------+-------------
Cached        |      -      |      9      |      -      |       -       |       -       |      -      |      -      |      1      |      14     |      24
Deleted       |      -      |      -      |      -      |       -       |       -       |      -      |      -      |      -      |      -      |      0
Duplicate     |     1070    |     2217    |     2290    |      293      |      1904     |     336     |     344     |      1      |      -      |     8455
Error         |      -      |      2      |      2      |       -       |       -       |      2      |      -      |      1      |      -      |      7
Ignored       |      -      |      -      |      -      |       -       |       -       |      -      |      -      |      9      |      -      |      9
Season-all    |      -      |      -      |      -      |       -       |       -       |      -      |      -      |      -      |      -      |      0
Skipped       |      94     |     2935    |     3394    |       -       |       -       |     264     |     279     |     1882    |      84     |     8932
Undefined     |      -      |      71     |      -      |       -       |       -       |      -      |      -      |      -      |      37     |     108
==============================================================================================================================================================
TOTAL         |     1164    |     5234    |     5686    |      293      |      1904     |     602     |     623     |     1894    |     135     |    17535
Download Time |      -      | 00:00:00.08 | 00:00:00.25 |       -       |       -       | 00:00:00.16 |      -      | 00:00:00.94 | 00:00:00.65 | 00:00:03.60

  Threads Used: 8
   Min/Avg/Max: 00.00 / 08.61 / 15.45 downloads per second
   Min/Avg/Max: 00.01 / 00.21 / 00.85 seconds per download

       Loading: 00:00:12.86
       Parsing: 00:00:00.30
     Comparing: 00:00:30.59
   Downloading: 00:00:03.60
 TOTAL RUNTIME: 00:00:47.37
```

**Comment by MilhouseVH** [comment]
`Duplicate` is for those items that are already cached AND appear in the metadata multiple times (eg. a tv show poster will appear at the `tvshow`, `season` and `episode` metadata levels ie. the same artwork is present potentially hundreds of times for one tv show)

`Skipped` is for an item of artwork that is already cached (ie present in the database) - this doesn't include `duplicate` items

As for the widget problem, my guess is your database `Textures13.db` and file items in the `Thumbnails` folder are not quite in sync.

You are most likely missing items in the file system (which is to say, you have database rows referencing artwork files that don't exist).

The `c` (cache) command is only concerned if an artwork item is present as a row in the database and it doesn't verify if the associated artwork item is also present in the file system (there's no Kodi API for that, so it would require `texturecache.py` having direct file system while caching which would make caching on remote clients impossible).

You can try running `texturecache.py X` directly on the Kodi client (not remotely as `texturecache.py` needs direct file system access) which will match each database row with the corresponding artwork item in the file system and then list those file items that are missing.

As long as the results look more-or-less correct/sane then you can use `texturecache.py Xd` which will automatically delete those database rows that don't have matching file items. Once you have synced the database and file system with `Xd` you can then re-cache (`c`) the database items you just removed - now you should have both rows and file system items in the cache.

Widgets are a bit out of scope for `texturecache.py` as they can only be cached by navigating the GUI (same for extracted thumbnails from tv show episodes). However if your database claims the widget is cached but the cached artwork is missing then it's possible that what you see for the widget in the GUI is incorrect (possibly a default artwork?). Synchronising your database and file system with `Xd` should resolve this kind of issue - you just need to get the GUI to re-cache the widget artwork.

Also, remember to check your `kodi.log` for errors when widgets are displayed incorrectly as it may give you some clues where things are going wrong.

**Comment by quadcom** [comment]
Thanks Mil,

In the end, I determined that my main system that I was copying the tc.db and the thumbs from had issues so it's no wonder the other systems had issues.

I ended up blowing away the entire DB and rescanning. I ran TC.py again and saw that there were 0 errors during the 'c' scan.

I guess at that point I could have copied over those files from that working system but instead deleted the thumbs and tc.db and rebooted the other systems and let them rebuild with TC.py. 

All systems are up and running and at the same time, I got a clearer understanding of how the thumbs system works along with rapidly restoring my library content by backing up the advancedsettigns.xml (and other shared XML's) before blowing away all sources. That saved a ton of time manually remapping them all.

Thanks for taking the time with a long and detailed writeup.
